File: Android/app/src/main/java/com/example/tavattendance/core/Analytics.kt
package com.example.tavattendance.core

import android.app.ActivityManager
import android.app.ApplicationExitInfo
import android.content.Context
import android.os.Build
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import com.example.tavattendance.data.service.FeatureFlags
import io.github.jan.supabase.postgrest.from
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.time.Instant
import java.util.UUID

/** Event type — mirrors the `app_events.event_type` CHECK (migration 031). */
enum class AnalyticsEventType(val raw: String) {
    SCREEN_VIEW("screen_view"), TAP("tap"), ERROR("error"),
    CRASH("crash"), OPS("ops"), LATENCY("latency")
}

/**
 * One `app_events` row. `properties` carries IDs/counts only — NEVER student
 * names (PDPA). Context fields (user/role/version/platform/device/session) are
 * stamped by [Analytics] at build time.
 */
@Serializable
data class AppEvent(
    @SerialName("occurred_at") val occurredAt: String,
    @SerialName("user_id") val userId: String?,
    val role: String?,
    val platform: String,
    @SerialName("app_version") val appVersion: String?,
    @SerialName("session_id") val sessionId: String,
    @SerialName("event_type") val eventType: String,
    val name: String,
    val properties: JsonObject,
    val device: String?
)

/**
 * Supabase-native, fail-silent product analytics + observability (migration 031).
 * Mirror of iOS `Services/Analytics.swift`.
 *
 * - No-op unless the `analytics` feature flag is ON (loaded once at sign-in;
 *   pre-flag events are dropped — acceptable).
 * - In-memory batch, flushed every 30s and when the app backgrounds.
 * - **Drop on failure**: a failed insert discards the batch. There is
 *   deliberately no second offline queue — attendance capture must never be
 *   blocked or complicated by analytics.
 */
object Analytics {
    private const val PLATFORM = "android"
    internal const val MAX_BUFFERED = 500
    private val EMPTY = JsonObject(emptyMap())

    private val lock = Any()
    /** Buffered events awaiting the next flush. Internal for tests. */
    internal val buffer = mutableListOf<AppEvent>()
    private val sessionId = UUID.randomUUID().toString()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    /** Set from `AuthViewModel` after the session/profile resolve. */
    var userId: String? = null
    var role: String? = null
    private var appVersion: String? = null
    private var device: String? = null
    private var started = false

    private val enabled: Boolean get() = FeatureFlags.isEnabled(FeatureFlags.ANALYTICS)

    // MARK: lifecycle

    /**
     * Stamps device/version context, checks for a crash since last launch, starts
     * the 30s flush loop and emits `app_launch`. Call once after flags load.
     */
    fun start(context: Context) {
        if (started) return
        started = true
        appVersion = runCatching {
            context.packageManager.getPackageInfo(context.packageName, 0).versionName
        }.getOrNull()
        device = Build.MODEL
        checkForCrash(context)
        scope.launch {
            while (isActive) {
                delay(30_000)
                flush()
            }
        }
        track(AnalyticsEventType.OPS, "app_launch", buildJsonObject { put("cold", true) })
    }

    // MARK: capture

    /** Public entry point — no-op while the flag is OFF. */
    fun track(type: AnalyticsEventType, name: String, properties: JsonObject = EMPTY) {
        if (!enabled) return
        record(type, name, properties)
    }

    /** Buffers unconditionally (no flag gate). The seam used by [track] and tests. */
    internal fun record(type: AnalyticsEventType, name: String, properties: JsonObject = EMPTY) {
        val event = AppEvent(
            occurredAt = Instant.now().toString(),
            userId = userId,
            role = role,
            platform = PLATFORM,
            appVersion = appVersion,
            sessionId = sessionId,
            eventType = type.raw,
            name = name,
            properties = properties,
            device = device
        )
        synchronized(lock) {
            buffer.add(event)
            // Long offline stretches must not grow the buffer without bound; shed oldest.
            if (buffer.size > MAX_BUFFERED) {
                repeat(buffer.size - MAX_BUFFERED) { buffer.removeAt(0) }
            }
        }
    }

    /** Handled-error funnel. `screen`/`message` are technical strings — never student names. */
    fun trackError(screen: String, error: Throwable) {
        track(AnalyticsEventType.ERROR, "handled_error", buildJsonObject {
            put("message", error.message ?: error.javaClass.simpleName)
            put("screen", screen)
        })
    }

    /** Fire-and-forget flush (app background / onStop). */
    fun flushNow() { scope.launch { flush() } }

    // MARK: flush

    /**
     * Sends the buffered batch, **dropping it on any failure** — the buffer is
     * cleared before the send, so a throwing [sink] simply discards those events.
     * [sink] is injectable for tests; production uses the Supabase insert.
     */
    internal suspend fun flush(sink: suspend (List<AppEvent>) -> Unit) {
        val batch: List<AppEvent>
        synchronized(lock) {
            if (buffer.isEmpty()) return
            batch = buffer.toList()
            buffer.clear()
        }
        runCatching { sink(batch) }
    }

    suspend fun flush() = flush { batch ->
        SupabaseClient.client.from("app_events").insert(batch)
    }

    // MARK: crash diagnostics (next-launch detection)

    private fun checkForCrash(context: Context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) return
        runCatching {
            val am = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
            val prefs = context.getSharedPreferences("analytics", Context.MODE_PRIVATE)
            val watermark = prefs.getLong("crash_watermark", 0L)
            var newest = watermark
            for (info in am.getHistoricalProcessExitReasons(null, 0, 20)) {
                if (info.timestamp <= watermark) continue
                newest = maxOf(newest, info.timestamp)
                val mechanism = when (info.reason) {
                    ApplicationExitInfo.REASON_CRASH -> "crash"
                    ApplicationExitInfo.REASON_CRASH_NATIVE -> "crash_native"
                    ApplicationExitInfo.REASON_ANR -> "anr"
                    else -> continue
                }
                track(AnalyticsEventType.CRASH, "crash_detected", buildJsonObject {
                    put("mechanism", mechanism)
                    put("reason", info.description ?: "unknown")
                })
            }
            prefs.edit().putLong("crash_watermark", newest).apply()
        }
    }
}

/** One line per staff screen — emits a `screen_view` on first composition. */
@Composable
fun TrackScreen(name: String) {
    LaunchedEffect(Unit) { Analytics.track(AnalyticsEventType.SCREEN_VIEW, name) }
}

File: iOS/TAVAttendance/Services/Analytics.swift
import Foundation
import SwiftUI
import UIKit
import MetricKit
import Supabase

/// Event type — mirrors the `app_events.event_type` CHECK (migration 031).
enum AnalyticsEventType: String {
    case screenView = "screen_view"
    case tap
    case error
    case crash
    case ops
    case latency
}

/// One `app_events` row. `properties` carries IDs/counts only — NEVER student
/// names (PDPA). Context fields (user/role/version/platform/device/session) are
/// stamped by `Analytics` at build time.
struct AppEvent: Encodable {
    let occurred_at: Date
    let user_id: String?
    let role: String?
    let platform: String
    let app_version: String?
    let session_id: String
    let event_type: String
    let name: String
    let properties: AnyJSON
    let device: String
}

/// Supabase-native, fail-silent product analytics + observability (migration 031).
///
/// - No-op unless the `analytics` feature flag is ON (loaded once at sign-in;
///   pre-flag events are dropped — acceptable).
/// - In-memory batch, flushed every 30s and when the app backgrounds.
/// - **Drop on failure** (`try?`): a failed insert discards the batch. There is
///   deliberately no second offline queue — attendance capture must never be
///   blocked or complicated by analytics.
@MainActor
final class Analytics: NSObject {
    static let shared = Analytics()

    /// Buffered events awaiting the next flush. Exposed for tests.
    private(set) var buffer: [AppEvent] = []

    /// Set once from `AuthManager` after the profile resolves.
    var role: String?

    private let db = SupabaseManager.shared.client
    private let sessionId = UUID().uuidString
    private let platform = "ios"
    private let appVersion = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
    private let device = UIDevice.current.model

    // Long offline stretches must not grow the buffer without bound; keep the
    // most recent events and shed the oldest.
    private let maxBuffered = 500

    private var flushTimer: Timer?
    private var started = false

    private var enabled: Bool { FeatureFlagStore.shared.isEnabled(.analytics) }

    // MARK: - Lifecycle

    /// Starts the flush timer, background-flush observer and MetricKit crash
    /// subscriber. Side-effect free until called (so tests can construct freely).
    func start() {
        guard !started else { return }
        started = true
        MXMetricManager.shared.add(self)

        NotificationCenter.default.addObserver(
            self, selector: #selector(handleBackground),
            name: UIApplication.didEnterBackgroundNotification, object: nil)

        flushTimer = Timer.scheduledTimer(withTimeInterval: 30, repeats: true) { [weak self] _ in
            Task { @MainActor in await self?.flush() }
        }

        track(.ops, name: "app_launch", properties: ["cold": .bool(true)])
    }

    @objc private func handleBackground() {
        Task { @MainActor in await flush() }
    }

    // MARK: - Capture

    /// Public entry point — no-op while the flag is OFF.
    func track(_ type: AnalyticsEventType, name: String, properties: JSONObject = [:]) {
        guard enabled else { return }
        record(type, name: name, properties: properties)
    }

    /// Buffers an event unconditionally (no flag gate). The flag gate lives in
    /// `track`; `record` is the seam used by tests and by `track`.
    func record(_ type: AnalyticsEventType, name: String, properties: JSONObject = [:]) {
        let event = AppEvent(
            occurred_at: Date(),
            user_id: db.auth.currentSession?.user.id.uuidString,
            role: role,
            platform: platform,
            app_version: appVersion,
            session_id: sessionId,
            event_type: type.rawValue,
            name: name,
            properties: .object(properties),
            device: device
        )
        buffer.append(event)
        if buffer.count > maxBuffered {
            buffer.removeFirst(buffer.count - maxBuffered)
        }
    }

    /// Times an async op and emits `{name}` with `duration_ms` folded into
    /// `properties` (merged with `extra`). Fires only on success; a throw
    /// propagates without an event (the error funnel captures the failure).
    @discardableResult
    func time<T>(_ name: String, type: AnalyticsEventType = .latency,
                 extra: JSONObject = [:], _ op: () async throws -> T) async rethrows -> T {
        let start = Date()
        let result = try await op()
        var props = extra
        props["duration_ms"] = .integer(Int(Date().timeIntervalSince(start) * 1000))
        track(type, name: name, properties: props)
        return result
    }

    /// Elapsed milliseconds since `start`, for events that also carry counts and
    /// so can't use the `time` wrapper.
    static func ms(since start: Date) -> AnyJSON { .integer(Int(Date().timeIntervalSince(start) * 1000)) }

    // MARK: - Flush

    /// Sends the buffered batch. **Drops the batch on any failure** — the buffer
    /// is cleared before the await, so a throw simply discards those events.
    /// `sink` is injectable for tests; production uses the Supabase insert.
    func flush(using sink: ([AppEvent]) async throws -> Void) async {
        guard !buffer.isEmpty else { return }
        let batch = buffer
        buffer.removeAll()
        try? await sink(batch)
    }

    func flush() async {
        await flush(using: { batch in
            struct Params: Encodable {
                let events: [AppEvent]
                enum CodingKeys: String, CodingKey { case events = "p_events" }
            }
            for start in stride(from: 0, to: batch.count, by: 100) {
                let end = min(start + 100, batch.count)
                try await self.db.rpc(
                    "submit_app_events",
                    params: Params(events: Array(batch[start..<end]))
                ).execute()
            }
        })
    }
}

// MARK: - MetricKit crash diagnostics (next-launch detection)

extension Analytics: MXMetricManagerSubscriber {
    nonisolated func didReceive(_ payloads: [MXMetricPayload]) {}

    nonisolated func didReceive(_ payloads: [MXDiagnosticPayload]) {
        let crashes: [(mechanism: String, reason: String)] = payloads.flatMap { payload in
            (payload.crashDiagnostics ?? []).map { d in
                (d.exceptionType.map { "exception:\($0)" } ?? "crash",
                 d.terminationReason ?? d.exceptionCode.map { "code:\($0)" } ?? "unknown")
            }
        }
        guard !crashes.isEmpty else { return }
        Task { @MainActor in
            for crash in crashes {
                track(.crash, name: "crash_detected",
                      properties: ["mechanism": .string(crash.mechanism), "reason": .string(crash.reason)])
            }
            await flush()
        }
    }
}

// MARK: - Screen-view modifier

extension View {
    /// One line per staff screen — emits a `screen_view` on appear.
    func analyticsScreen(_ name: String) -> some View {
        onAppear { Analytics.shared.track(.screenView, name: name) }
    }
}
